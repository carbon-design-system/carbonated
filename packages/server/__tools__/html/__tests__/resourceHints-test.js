/**
 * @jest-environment node
 */

'use strict';

const path = require('path');

describe('resourceHints', () => {
  let assetManifest;
  let resourceHints;
  let preloadHints;
  let prefetchHints;
  let print;

  beforeEach(() => {
    assetManifest = require('../__fixtures__/asset-manifest.json');
    resourceHints = require('../resourceHints');
    preloadHints = resourceHints.preloadHints;
    prefetchHints = resourceHints.prefetchHints;
    print = resourceHints.print;
  });

  describe('#defaultExtract', () => {
    it('should remove sourceMaps from the asset manifest', () => {
      const extracted = preloadHints.extract(assetManifest);
      extracted.forEach(resourceHint => {
        expect(path.extname(resourceHint.chunkName)).not.toBe('.map');
      });
    });

    it('should remove any blacklisted elements from the asset manifest', () => {
      const { resourceBlacklist } = resourceHints;
      const extracted = preloadHints.extract(assetManifest);
      resourceBlacklist.forEach(resourceName => {
        const resource = extracted.find(resourceHint => {
          return resourceHint.chunkName.indexOf(resourceName) !== -1;
        });
        expect(resource).not.toBeDefined();
      });
    });
  });

  describe('#preloadHints', () => {
    it('should define an extract method', () => {
      expect(preloadHints.extract).toBeDefined();
    });

    it('should extract chunks defined in `preloadChunks` as resource hints', () => {
      const { preloadChunks, resourceBlacklist } = resourceHints;
      const extracted = preloadHints.extract(assetManifest);

      preloadChunks.forEach(chunkName => {
        const resource = extracted.find(resourceHint => {
          return resourceHint.chunkName.indexOf(chunkName) !== -1;
        });
        expect(resource).toBeDefined();
      });

      resourceBlacklist.forEach(resourceName => {
        const resource = extracted.find(resourceHint => {
          return resourceHint.chunkName.indexOf(resourceName) !== -1;
        });
        expect(resource).not.toBeDefined();
      });
    });

    it('should print out link tags', () => {
      const { extract } = preloadHints;
      const tags = print(extract(assetManifest));
      expect(tags).toMatchSnapshot();
    });
  });

  describe('#prefetchHints', () => {
    it('should define an extract method', () => {
      expect(prefetchHints.extract).toBeDefined();
    });

    it('should extract non-preload chunks', () => {
      const { preloadChunks, resourceBlacklist } = resourceHints;
      const extracted = prefetchHints.extract(assetManifest);

      preloadChunks.forEach(chunkName => {
        const resource = extracted.find(resourceHint => {
          return resourceHint.chunkName.indexOf(chunkName) !== -1;
        });
        expect(resource).not.toBeDefined();
      });

      resourceBlacklist.forEach(resourceName => {
        const resource = extracted.find(resourceHint => {
          return resourceHint.chunkName.indexOf(resourceName) !== -1;
        });
        expect(resource).not.toBeDefined();
      });
    });

    it('should print out link tags', () => {
      const { extract } = prefetchHints;
      const tags = print(extract(assetManifest));
      expect(tags).toMatchSnapshot();
    });
  });
});
