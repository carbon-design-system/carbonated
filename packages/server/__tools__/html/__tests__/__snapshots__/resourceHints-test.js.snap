// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resourceHints #prefetchHints should print out link tags 1`] = `
Array [
  "<link rel=\\"prefetch\\" href=\\"/static/js/Foo.d32c4417.chunk.js\\" as=\\"script\\">",
]
`;

exports[`resourceHints #preloadHints should print out link tags 1`] = `
Array [
  "<link rel=\\"preload\\" href=\\"static/css/main.1c0d3fed.css\\" as=\\"style\\">",
  "<link rel=\\"preload\\" href=\\"static/js/main.85d0f27c.js\\" as=\\"script\\">",
  "<link rel=\\"preload\\" href=\\"static/js/vendor.7f3d1409.js\\" as=\\"script\\">",
]
`;
